packages/ckeditor5-ui/src/editorui/bodycollection.ts

```diff
@@ -82,14 +88,12 @@ export default class BodyCollection extends ViewCollection {
 			children: this
 		} ).render() as HTMLElement;
 
-		let wrapper = document.querySelector( '.ck-body-wrapper' );
-
-		if ( !wrapper ) {
-			wrapper = createElement( document, 'div', { class: 'ck-body-wrapper' } );
-			document.body.appendChild( wrapper );
+		if ( !BodyCollection._bodyWrapper || !BodyCollection._bodyWrapper.isConnected ) {
```

Comment:
I originally added the `|| !BodyCollection._bodyWrapper.isConnected` to fix the tests before I figured out to add [this line](https://github.com/ckeditor/ckeditor5/pull/17321/files#diff-f600513e49b56cb5c923a1518409d6c2ce166751f21e603ec7b3435e5540575cR29). So this isn't strictly needed, but seems like a good safety mechanism so that if somebody somehow removes the body wrapper from the DOM, `BodyCollection`s aren't forever broken, installing themselves in a detached element.

Reply:
This is what I had in mind but I'd keep it as simple as possible. 

With the `isConnected` condition that you proposed, if somebody removed the wrapper from DOM, its reference would be lost when a new editor instance gets created (also another wrapper would be created). If the old wrapper somehow gets back into DOM (e.g. the same way it got lost in the first place), it will remain an orphan forever even if you destroy all editor instances. And there will be two or more wrappers in DOM at a time, which itself is quite unpredictable.

In such a scenario, I'd rather see the UI of all editors broken: at least it would be obvious that something is wrong. If the integrator decides to manage `.ck-body` on their own, it's their job to ensure it sits in DOM when new editor instances are created.

I know you could make this logic even more bulletproof, e.g. store references to all (past) wrappers, iterate over them and so on. But this will be nothing more than a workaround for a problem that should not exist in the first place: there should be one `.ck-body` wrapper at a time that sits in DOM all the time.

---

P.S. We'll need [tests for the changes to accept this PR](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/contributing.html#tests). Thank you!

Reply:
Okay, so remove the `|| !BodyCollection._bodyWrapper.isConnected` -- no problem.

As far as testing, I'm not exactly sure what kind of test to write since the end-result behavior (the presence and singleton-ness of the body wrapper) is already tested. I could add an `expect( wrappers[ 0 ] ).to.equal( BodyCollection._bodyWrapper );` [here](https://github.com/ckeditor/ckeditor5/blob/f70eedd6879dc0a0e6a1183381d494f0b7c0ab57/packages/ckeditor5-ui/tests/editorui/bodycollection.js#L57) and/or [here](https://github.com/ckeditor/ckeditor5/blob/f70eedd6879dc0a0e6a1183381d494f0b7c0ab57/packages/ckeditor5-ui/tests/editorui/bodycollection.js#L112). Does that make sense? Or do you think we need a new test, and if so, what is it testing?